// @genkit-ai/firebase/lib/index.mjs
import {
  __async
} from "./chunk-DJRN6NKF.mjs";
import { genkitPlugin } from "@genkit-ai/core";
import { FirestoreStateStore } from "@genkit-ai/flow";
import {
  GcpLogger,
  GcpOpenTelemetry
} from "@genkit-ai/google-cloud";
import { GoogleAuth } from "google-auth-library";
import { FirestoreTraceStore } from "./firestoreTraceStore.js";
import { defineFirestoreRetriever } from "./firestoreRetriever.js";
const firebase = genkitPlugin(
  "firebase",
  (params) => __async(void 0, null, function* () {
    const authClient = new GoogleAuth();
    const gcpOptions = {
      projectId: (params == null ? void 0 : params.projectId) || (yield authClient.getProjectId()),
      telemetryConfig: params == null ? void 0 : params.telemetryConfig
    };
    return {
      flowStateStore: {
        id: "firestore",
        value: new FirestoreStateStore(params == null ? void 0 : params.flowStateStore)
      },
      traceStore: {
        id: "firestore",
        value: new FirestoreTraceStore(params == null ? void 0 : params.traceStore)
      },
      telemetry: {
        instrumentation: {
          id: "firebase",
          value: new GcpOpenTelemetry(gcpOptions)
        },
        logger: {
          id: "firebase",
          value: new GcpLogger(gcpOptions)
        }
      }
    };
  })
);
export {
  defineFirestoreRetriever,
  firebase
};
