// @genkit-ai/google-cloud/lib/gcpOpenTelemetry.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var gcpOpenTelemetry_exports = {};
__export(gcpOpenTelemetry_exports, {
  GcpOpenTelemetry: () => GcpOpenTelemetry,
  __forceFlushSpansForTesting: () => __forceFlushSpansForTesting,
  __getMetricExporterForTesting: () => __getMetricExporterForTesting,
  __getSpanExporterForTesting: () => __getSpanExporterForTesting
});
module.exports = __toCommonJS(gcpOpenTelemetry_exports);
var import_opentelemetry_cloud_monitoring_exporter = require("@google-cloud/opentelemetry-cloud-monitoring-exporter");
var import_opentelemetry_cloud_trace_exporter = require("@google-cloud/opentelemetry-cloud-trace-exporter");
var import_opentelemetry_resource_util = require("@google-cloud/opentelemetry-resource-util");
var import_api = require("@opentelemetry/api");
var import_auto_instrumentations_node = require("@opentelemetry/auto-instrumentations-node");
var import_instrumentation_pino = require("@opentelemetry/instrumentation-pino");
var import_instrumentation_winston = require("@opentelemetry/instrumentation-winston");
var import_resources = require("@opentelemetry/resources");
var import_sdk_metrics = require("@opentelemetry/sdk-metrics");
var import_sdk_trace_base = require("@opentelemetry/sdk-trace-base");
let metricExporter;
let spanProcessor;
let spanExporter;
class GcpOpenTelemetry {
  constructor(options) {
    /**
     * Log hook for writing trace and span metadata to log messages in the format
     * required by GCP.
     */
    this.gcpTraceLogHook = (span, record) => {
      const isSampled = !!(span.spanContext().traceFlags & import_api.TraceFlags.SAMPLED);
      record["logging.googleapis.com/trace"] = `projects/${this.options.projectId}/traces/${span.spanContext().traceId}`;
      record["logging.googleapis.com/spanId"] = span.spanContext().spanId;
      record["logging.googleapis.com/trace_sampled"] = isSampled ? "1" : "0";
    };
    this.options = options || {};
    this.resource = new import_resources.Resource({ type: "global" }).merge(
      new import_opentelemetry_resource_util.GcpDetectorSync().detect()
    );
  }
  getConfig() {
    var _a, _b;
    spanProcessor = new import_sdk_trace_base.BatchSpanProcessor(this.createSpanExporter());
    return {
      resource: this.resource,
      spanProcessor,
      sampler: ((_b = (_a = this.options) == null ? void 0 : _a.telemetryConfig) == null ? void 0 : _b.sampler) || new import_sdk_trace_base.AlwaysOnSampler(),
      instrumentations: this.getInstrumentations(),
      metricReader: this.createMetricReader()
    };
  }
  createSpanExporter() {
    spanExporter = new AdjustingTraceExporter(
      this.shouldExportTraces() ? new import_opentelemetry_cloud_trace_exporter.TraceExporter() : new import_sdk_trace_base.InMemorySpanExporter()
    );
    return spanExporter;
  }
  /**
   * Creates a {MetricReader} for pushing metrics out to GCP via OpenTelemetry.
   */
  createMetricReader() {
    var _a, _b, _c, _d;
    metricExporter = this.buildMetricExporter();
    return new import_sdk_metrics.PeriodicExportingMetricReader({
      exportIntervalMillis: ((_b = (_a = this.options) == null ? void 0 : _a.telemetryConfig) == null ? void 0 : _b.metricExportIntervalMillis) || 3e5,
      exportTimeoutMillis: ((_d = (_c = this.options) == null ? void 0 : _c.telemetryConfig) == null ? void 0 : _d.metricExportTimeoutMillis) || 3e5,
      exporter: metricExporter
    });
  }
  /** Gets all open telemetry instrumentations as configured by the plugin. */
  getInstrumentations() {
    var _a, _b, _c, _d;
    if ((_b = (_a = this.options) == null ? void 0 : _a.telemetryConfig) == null ? void 0 : _b.autoInstrumentation) {
      return (0, import_auto_instrumentations_node.getNodeAutoInstrumentations)(
        ((_d = (_c = this.options) == null ? void 0 : _c.telemetryConfig) == null ? void 0 : _d.autoInstrumentationConfig) || {}
      ).concat(this.getDefaultLoggingInstrumentations());
    }
    return this.getDefaultLoggingInstrumentations();
  }
  shouldExportTraces() {
    var _a, _b;
    return (((_a = this.options.telemetryConfig) == null ? void 0 : _a.forceDevExport) || process.env.GENKIT_ENV !== "dev") && !((_b = this.options.telemetryConfig) == null ? void 0 : _b.disableTraces);
  }
  shouldExportMetrics() {
    var _a, _b;
    return (((_a = this.options.telemetryConfig) == null ? void 0 : _a.forceDevExport) || process.env.GENKIT_ENV !== "dev") && !((_b = this.options.telemetryConfig) == null ? void 0 : _b.disableMetrics);
  }
  /** Always configure the Pino and Winston instrumentations */
  getDefaultLoggingInstrumentations() {
    return [
      new import_instrumentation_winston.WinstonInstrumentation({ logHook: this.gcpTraceLogHook }),
      new import_instrumentation_pino.PinoInstrumentation({ logHook: this.gcpTraceLogHook })
    ];
  }
  buildMetricExporter() {
    const exporter = this.shouldExportMetrics() ? new import_opentelemetry_cloud_monitoring_exporter.MetricExporter({ projectId: this.options.projectId }) : new import_sdk_metrics.InMemoryMetricExporter(import_sdk_metrics.AggregationTemporality.DELTA);
    exporter.selectAggregation = (instrumentType) => {
      if (instrumentType === import_sdk_metrics.InstrumentType.HISTOGRAM) {
        return new import_sdk_metrics.ExponentialHistogramAggregation();
      }
      return new import_sdk_metrics.DefaultAggregation();
    };
    exporter.selectAggregationTemporality = (instrumentType) => {
      return import_sdk_metrics.AggregationTemporality.DELTA;
    };
    return exporter;
  }
}
class AdjustingTraceExporter {
  constructor(exporter) {
    this.exporter = exporter;
  }
  export(spans, resultCallback) {
    var _a;
    (_a = this.exporter) == null ? void 0 : _a.export(this.adjust(spans), resultCallback);
  }
  shutdown() {
    var _a;
    return (_a = this.exporter) == null ? void 0 : _a.shutdown();
  }
  getExporter() {
    return this.exporter;
  }
  forceFlush() {
    var _a;
    if ((_a = this.exporter) == null ? void 0 : _a.forceFlush) {
      return this.exporter.forceFlush();
    }
    return Promise.resolve();
  }
  adjust(spans) {
    return spans.map((span) => {
      span = this.redactPii(span);
      span = this.markErrorSpanAsError(span);
      span = this.normalizeLabels(span);
      return span;
    });
  }
  redactPii(span) {
    const hasInput = "genkit:input" in span.attributes;
    const hasOutput = "genkit:output" in span.attributes;
    return !hasInput && !hasOutput ? span : __spreadProps(__spreadValues({}, span), {
      spanContext: span.spanContext,
      attributes: __spreadProps(__spreadValues({}, span.attributes), {
        "genkit:input": "<redacted>",
        "genkit:output": "<redacted>"
      })
    });
  }
  // This is a workaround for GCP Trace to mark a span with a red
  // exclamation mark indicating that it is an error.
  markErrorSpanAsError(span) {
    return span.status.code !== import_api.SpanStatusCode.ERROR ? span : __spreadProps(__spreadValues({}, span), {
      spanContext: span.spanContext,
      attributes: __spreadProps(__spreadValues({}, span.attributes), {
        "/http/status_code": "599"
      })
    });
  }
  // This is a workaround for GCP Trace to mark a span with a red
  // exclamation mark indicating that it is an error.
  normalizeLabels(span) {
    const normalized = {};
    for (const [key, value] of Object.entries(span.attributes)) {
      normalized[key.replace(/\:/g, "/")] = value;
    }
    return __spreadProps(__spreadValues({}, span), {
      spanContext: span.spanContext,
      attributes: normalized
    });
  }
}
function __getMetricExporterForTesting() {
  return metricExporter;
}
function __getSpanExporterForTesting() {
  return spanExporter.getExporter();
}
function __forceFlushSpansForTesting() {
  spanProcessor.forceFlush();
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  GcpOpenTelemetry,
  __forceFlushSpansForTesting,
  __getMetricExporterForTesting,
  __getSpanExporterForTesting
});
